### PR TITLE
Feat: Edit Commands

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
When users edit the message in their text commands, the text command processor will now (configurably) execute it.

# Details
By setting `TextCommandsConfiguration.EnableEditCommands` to `true`, the message update event will now cause commands to execute.

> plus making sure the regular context type filtering behaviour actually works with the new feature

@akiraveliara mentioned wanting to introduce a new context type for this (presumably `EditedTextCommandContext`), which I'm in full support of. However I no longer have the time to implement that much nor do I have any idea of the full feature set she wants to bring to the table. Hopefully this PR can serve as a base for it.

Requesting @VelvetToroyashi's review for the message flags specifically.

# Notes
Tested.

### CODE REVIEW

File: DSharpPlus.Commands/Processors/TextCommands/TextCommandProcessor.cs [hunk 1]
```diff
@@ -64,15 +67,64 @@ public override async ValueTask ConfigureAsync(CommandsExtension extension)
...
+            SlidingExpiration = TimeSpan.FromSeconds(15)
+        });
+
+        // Check to see if the flags are the same, excluding the flags that are not relevant to the message content.
+        DiscordMessageFlags notTheseFlags = DiscordMessageFlags.Crossposted | DiscordMessageFlags.IsCrosspost | DiscordMessageFlags.SuppressedEmbeds;
+
+        // Flags are null when 0. 0 is not equal to null.
+        DiscordMessageFlags relevantFlags = eventArgs.Message.Flags & ~notTheseFlags ?? 0;
+        DiscordMessageFlags relevantFlagsBefore = eventArgs.MessageBefore?.Flags & ~notTheseFlags ?? 0;
+        if (eventArgs.MessageBefore is not null && relevantFlags == relevantFlagsBefore)
```
Comment (OoLunar) [inline DSharpPlus.Commands/Processors/TextCommands/TextCommandProcessor.cs:102]: Surely there's a better way to do this?

Comment (VelvetToroyashi) [inline DSharpPlus.Commands/Processors/TextCommands/TextCommandProcessor.cs:102]: Checking flags is fine but I would also check if the content is actually updated. If this is irrelevant please note I'm on mobile and will do a more thorough look through soon

Comment (OoLunar) [reply DSharpPlus.Commands/Processors/TextCommands/TextCommandProcessor.cs:102]: No this is right actually. Good catch

File: DSharpPlus.Commands/ProcessorInvokingHandlers.cs [hunk 2]
```diff
@@ -1,50 +1,59 @@
...
-
-    public async Task HandleEventAsync(DiscordClient sender, MessageCreatedEventArgs eventArgs)
-    {
-        if (this.extension.TryGetProcessor(out TextCommandProcessor? processor))
-        {
-            await processor.ExecuteTextCommandAsync(sender, eventArgs);
-        }
-    }
-}
+using System.Threading.Tasks;
```
Comment (VelvetToroyashi) [inline DSharpPlus.Commands/ProcessorInvokingHandlers.cs:1]: One day GitHub will learn how to Diff properly

Comment (VelvetToroyashi) [reply DSharpPlus.Commands/ProcessorInvokingHandlers.cs:1]: (Actually on second thought, your whitespace is botched here - line endings probably?)

**Review by VelvetToroyashi — APPROVED** [review]
Not sure how I feel about the message cache, but looks fine

File: DSharpPlus.Commands/ProcessorInvokingHandlers.cs (hunk 2 above)
Comment (OoLunar) [reply DSharpPlus.Commands/ProcessorInvokingHandlers.cs:1]: We should bring the `.gitattributes` file back

**Review by akiraveliara — CHANGES_REQUESTED** [review]
1. line endings in ProcessorInvokingHandlers.cs
2. i'm not sure we actually need the message cache - just checking whether the current time is within a second or thereabouts of EditedTimestamp should be fine, barring disastrous edge-cases
3. commands should be told whether their context is from an edit or not.

**Comment by OoLunar** [comment]
> commands should be told whether their context is from an edit or not.

As stated in the PR description, I have no plans on implementing this. You may extend or commit directly to my PR, however there is no implementation I can currently come up with that I would be satisfied with. Ideally the `EditedTextCommandContext` class would only have a single property: `MessageBefore` and there would be a `EditedTextConverterContext` available for argument converters to use as well.

A possible solution would involve modulating all of the code within `ExecuteTextCommandAsync`, similar to how the `DefaultCommandExecutor` works. As command processors were intended to be small when originally designing the extension, it's entirely possible for there to be a `EditedTextCommandProcessor`, or something of similar naming. Alternatively we could just keep adding onto the current text command processor, since it's quite bulky already.

**Comment by akiraveliara** [comment]
closing as this not a PR we are currently interested in taking and has not seen any activity

**Pull request closed without merging**